### ISSUE TITLE
State changes are not propagated properly across epochs

### ISSUE THREAD

**Issue by lacabra** [enigmampc/discovery-docker-network#21]
**Describe the bug**
Fails testing a secret contract that modifies the state, and relies on the state change properly propagating across epochs and workers.

When the epoch changes, the worker that gets assigned to the second computations throws the following error:
```
core_3      | building state for 1 contracts
core_3      | Failed applying delta: SystemError(StateError { err: "Hashes don\'t match, Failed Applying the delta" })
core_3      | 19:34:40 [INFO] ptt_build_state() => [Hash256([136, 152, 122, 247, 211, 94, 171, 202, 217, 89, 21, 185, 59, 253, 61, 43, 195, 48, 143, 6, 183, 25, 116, 120, 176, 223, 202, 38, 143, 4, 151, 220])]
core_3      | 19:34:40 [INFO] ptt_response() => Ok(PTTResponse { result: Errors([IpcStatusResult { address: "88987af7d35eabcad95915b93bfd3d2bc3308f06b7197478b0dfca268f0497dc", key: None, status: -1 }]) })
p2p_3       | [Tue May 28 2019 19:34:40 GMT+0000 (Coordinated Universal Time)] ERROR Failed Core connection: err: [object Object], coreResponse: {"id":"KY3nllUpLW8k","type":"PTTResponse","result":{"errors":[{"address":"88987af7d35eabcad95915b93bfd3d2bc3308f06b7197478b0dfca268f0497dc", "status":-1}]}}
```

and when that worker receives the Computation task, it errors out with the following:
```
[INFO] compute_task() => Err(DBErr { command: "read", kind: MissingKey }
```

**To Reproduce**
1. Launch the network with 3 nodes: `NODES=3 ./launch.bash`
2. Execute the following tests in this order:
```
01_init.spec.js 
02_deploy_millionaire.spec.js
10_execute_millionaire.spec.js
```
where the third one requires the `contract` repo to use the `stateacrossepochs` branch, as per enigmampc/enigma-contract#104

**Expected behavior**
The integrationTest should pass, and work the same it works when there is only 1 node in the network, or when all the computations happen within the same epoch with multiple nodes.

**Comment by AvishaiW** [comment]
@lacabra I'm looking into this and I'm noticing the following behavior: worker_1 deploys the contract C and updates the delta and state on his local db, then an epoch changes where worker_2 is in charge of contract C so he needs to build the state on his side. he gets a delta (via the IPC command: update_delttas), however, I didn't notice that worker_1 at any point sent his local deltas. I'll look into this a bit more both on core and on the p2p side tomorrow with @lenak25 to locate the exact location it's failing.

**Comment by AvishaiW** [comment]
after a bit more research with @lenak25 we noticed that when a contract is deployed, the worker broadcasts only the contract to the network and ignores the first delta, leading to a defect when other workers try to build the state- this was notified due to the change of the epoch.
in order to not overload the network, we will fix it by adding a new ipc message which is dedicated specifically to deployments which will contain both the bytecode and the delta. 
Thanks for catching this!

**Comment by AvishaiW** [comment]
@lacabra the changes both on core and on p2p were pushed to the develop branch, however, when trying to run the tests, during the execution `10_execute_millionaire.spec.js` we get the following error for each of the workers:
```
core_2      | 14:29:32 [INFO] get_ptt_req() => Ok(GetPTTRequest { result: Request { request: "83a46461746181a752657175657374c0a269649cccaaccc74b6fccab6a011fccbf1f57ccf9a67075626b6579dc0040ccb5ccd9ccba4acca2cca0355e41ccc5cca802ccd7cceecce0ccd22049cc890ecccc511e7bcca524cc9156cc9921ccd877515f3127cccbccc9594041cced6f21143449ccd73acc9c0076ccbc2f6956ccf4431bccff6a32ccb76a", sig: "e374ea4a20422495c51c980ede122abbb0a4e9b2a46d4e83815af0b81d0800dd60c3f4210a235a82d3da13caa5f07c6077da842b5f6cbf1b48203ce850283a801c" } })
km_1        | Parsed the WorkersParameterized event: Log { params: [LogParam { name: "seed", value: Uint(63895909724510537833854465854776920952101192335073547155735530052751868132938) }, LogParam { name: "firstBlockNumber", value: Uint(65) }, LogParam { name: "inclusionBlockNumber", value: Uint(64) }, LogParam { name: "workers", value: Array([Address(0x8f029cef914044ce194126b309e4398e04784017), Address(0x558df06602379d54a25a0b11d8e5a0b246117afc), Address(0xce394bd941f46a1215829fefc3ffa119bad016ee)]) }, LogParam { name: "stakes", value: Array([Uint(90000000000), Uint(90000000000), Uint(90000000000)]) }, LogParam { name: "nonce", value: Uint(3) }] }
km_1        | Got get_state_keys request: StateKeyRequest { data: StringWrapper("83a46461746181a752657175657374c0a269649ccca87bccbc047fcc964b7b7240cc8fcc8ba67075626b6579dc0040680e0d132036cca855cce5ccaacc9ecca6cca77bccb8cccacceecccb31ccd2ccaa39cca63ecc8376ccf04e53223d6bcca6cccc74ccb3ccfb75ccdcccbe18ccdcccfaccaccc8dccf21f422ccc9ccccccc81cc88cce9cc86321016ccba58cccaccd4cca703"), sig: StringWrapper("bf2f90ee54c409d049eb6a473abb74f4b3098cf5ac4cb3d93cba1eb23834405d104918c39d6d7fe62989460483b51d5bcaad853f1f0c2e2166548bd42b5342581c"), block_number: None }
km_1        | Got get_state_keys request: StateKeyRequest { data: StringWrapper("83a46461746181a752657175657374c0a269649cccaaccc74b6fccab6a011fccbf1f57ccf9a67075626b6579dc0040ccb5ccd9ccba4acca2cca0355e41ccc5cca802ccd7cceecce0ccd22049cc890ecccc511e7bcca524cc9156cc9921ccd877515f3127cccbccc9594041cced6f21143449ccd73acc9c0076ccbc2f6956ccf4431bccff6a32ccb76a"), sig: StringWrapper("e374ea4a20422495c51c980ede122abbb0a4e9b2a46d4e83815af0b81d0800dd60c3f4210a235a82d3da13caa5f07c6077da842b5f6cbf1b48203ce850283a801c"), block_number: None }
p2p_2       | [Thu Jun 06 2019 14:29:32 GMT+0000 (Coordinated Universal Time)] DEBUG Connecting to principal node: http://172.19.0.4:3040
p2p_2       | [Thu Jun 06 2019 14:29:32 GMT+0000 (Coordinated Universal Time)] ERROR Failed Principal node connection: -32603 - Unable to get keys: Operation not allowed while the EpochState is transitioning. Current state = UNCONFIRMED
```

**Comment by lenak25** [comment]
@lacabra , any update on this?

**Comment by lacabra** [comment]
@lenak25 and @AvishaiW: with the latest `develop` branches from all the repos, the unrelated error that @AvishaiW reported above is now gone, so we can go back to testing the original error, which is still present in a slightly different form. Now `core` builds the state correctly and runs the computation, but are the other `p2p` instances which fail to verify the task with the following error:
```
p2p_3       | [Wed Jun 26 2019 16:33:26 GMT+0000 (Coordinated Universal Time)] INFO [VERIFY_TASK_RESULT] error in verification of result of task 0xfc8a4d0eef0a61b9ba28325bb5caeb20be76883ce82a52b98c8a88c243258cc3: TaskVerificationErr: Mismatch in delta hash in task result 0xfc8a4d0eef0a61b9ba28325bb5caeb20be76883ce82a52b98c8a88c243258cc3
p2p_3       | [Wed Jun 26 2019 16:33:26 GMT+0000 (Coordinated Universal Time)] DEBUG [UPDATE_DB] : is_err ?  TaskVerificationErr: Mismatch in delta hash in task result 0xfc8a4d0eef0a61b9ba28325bb5caeb20be76883ce82a52b98c8a88c243258cc3
p2p_1       | [Wed Jun 26 2019 16:33:26 GMT+0000 (Coordinated Universal Time)] INFO [VERIFY_TASK_RESULT] error in verification of result of task 0xfc8a4d0eef0a61b9ba28325bb5caeb20be76883ce82a52b98c8a88c243258cc3: TaskVerificationErr: Mismatch in delta hash in task result 0xfc8a4d0eef0a61b9ba28325bb5caeb20be76883ce82a52b98c8a88c243258cc3
p2p_1       | [Wed Jun 26 2019 16:33:26 GMT+0000 (Coordinated Universal Time)] DEBUG [UPDATE_DB] : is_err ?  TaskVerificationErr: Mismatch in delta hash in task result 0xfc8a4d0eef0a61b9ba28325bb5caeb20be76883ce82a52b98c8a88c243258cc3
```
The above error can be reproduced in the very same way as the original error was triggered, namely:
1. Launch the network with 3 nodes: NODES=3 ./launch.bash
2. Execute the following tests in this order:
```
01_init.spec.js 
02_deploy_millionaire.spec.js
10_execute_millionaire.spec.js
```
I am able to reproduce the above error with all repos on the `develop` branch, without having to switch the `contract` to `stateacrossepochs`

**Comment by lenak25** [comment]
@lacabra , I was able to reproduce the issue locally and to understand its root cause: a small inconsistency between core and P2P on the definition of _No Delta_ (while P2P expects to have no **data** field, core return an empty field). @AvishaiW and I will address it.
Thanks for raising this.

**Comment by lacabra** [comment]
Fixed via enigmampc/enigma-p2p#212 and tested via enigmampc/enigma-contract#104
